File: plugin/src/main/java/ru/arc/xserver/JoinMessages.java
package ru.arc.xserver;



import lombok.Getter;
import ru.arc.xserver.repos.RepoData;

import java.util.HashSet;
import java.util.Set;

@Getter
public class JoinMessages extends RepoData<JoinMessages> {

    String player;
    Set<String> joinMessages = new HashSet<>();
    Set<String> leaveMessages = new HashSet<>();
    long timestamp = System.currentTimeMillis();

    public JoinMessages(String player) {
        this.player = player;
        this.timestamp = System.currentTimeMillis();
    }

    public void addJoinMessage(String message) {
        joinMessages.add(message);
        setDirty(true);
    }

    public void removeJoinMessage(String message) {
        joinMessages.remove(message);
        setDirty(true);
    }

    public void addLeaveMessage(String message) {
        leaveMessages.add(message);
        setDirty(true);
    }

    public void removeLeaveMessage(String message) {
        leaveMessages.remove(message);
        setDirty(true);
    }

    @Override
    public String id() {
        return player;
    }

    @Override
    public boolean isRemove() {
        return System.currentTimeMillis() - timestamp > 1000 * 60 * 60 * 24 * 7 && joinMessages.isEmpty() && leaveMessages.isEmpty();
    }

    @Override
    public void merge(JoinMessages other) {
        joinMessages.clear();
        joinMessages.addAll(other.joinMessages);
        leaveMessages.clear();
        leaveMessages.addAll(other.leaveMessages);
    }
}

File: src/main/kotlin/ru/arc/xserver/JoinMessages.kt
package ru.arc.xserver

import ru.arc.join.JoinAnnouncementKind
import ru.arc.repository.Entity
import ru.arc.repository.Mergeable
import java.util.concurrent.ThreadLocalRandom

class JoinMessages(
    @JvmField var player: String,
) : Entity,
    Mergeable<JoinMessages> {
    constructor() : this("")

    @Volatile
    @JvmField
    var joinMessages: Set<String> = emptySet()

    @Volatile
    @JvmField
    var leaveMessages: Set<String> = emptySet()

    @Volatile
    @JvmField
    var customJoinMessages: Set<String> = emptySet()

    @Volatile
    @JvmField
    var customLeaveMessages: Set<String> = emptySet()

    @JvmField
    var timestamp: Long = System.currentTimeMillis()

    init {
        timestamp = System.currentTimeMillis()
    }

    fun randomJoinMessage(): String? = randomFrom(joinMessages)

    fun randomLeaveMessage(): String? = randomFrom(leaveMessages)

    fun randomMessage(kind: JoinAnnouncementKind): String? =
        when (kind) {
            JoinAnnouncementKind.FIRST_TIME -> null
            JoinAnnouncementKind.JOIN -> randomFrom(joinMessages)
            JoinAnnouncementKind.LEAVE -> randomFrom(leaveMessages)
        }

    override fun id(): String = player

    @Synchronized
    override fun merge(other: JoinMessages) {
        joinMessages = other.joinMessages.orEmpty().toSet()
        leaveMessages = other.leaveMessages.orEmpty().toSet()
        customJoinMessages = other.customJoinMessages.orEmpty().toSet()
        customLeaveMessages = other.customLeaveMessages.orEmpty().toSet()
        timestamp = other.timestamp
    }

    companion object {
        const val CUSTOM_MESSAGE_PREFIX = "%player_name% "
        const val CUSTOM_TEMPLATE_PREFIX = CustomJoinMessageTemplate.FULL_PREFIX
        const val MAX_CUSTOM_MESSAGES_PER_KIND = 10
        const val MAX_CUSTOM_MESSAGE_LENGTH = 120
        const val MAX_CUSTOM_TEMPLATE_LENGTH = CustomJoinMessageTemplate.MAX_LENGTH
        const val MAX_CUSTOM_VISIBLE_LENGTH = CustomJoinMessageTemplate.MAX_VISIBLE_LENGTH

        fun validCustomMessages(messages: Set<String>): List<String> =
            messages.asSequence()
                .filter(CustomJoinMessageTemplate::valid)
                .map(String::trim)
                .distinct()
                .take(MAX_CUSTOM_MESSAGES_PER_KIND)
                .toList()

        fun customSelectionKey(message: String): String = CustomJoinMessageTemplate.selectionKey(message.trim())
    }

    private fun randomFrom(messages: Set<String>): String? {
        val snapshot = messages.filter(String::isNotBlank)
        if (snapshot.isEmpty()) return null
        return snapshot[ThreadLocalRandom.current().nextInt(snapshot.size)]
    }
}
